src/test/HST/Util/SubstTests.hs

```diff
@@ -61,124 +56,310 @@ currentFrontendSyntax = return
+      $
+      -- σ    = { z ↦ x }
+      -- e    = case x of
+      --          x -> z
+      --           where z = x
+      --          y -> z
+      -- σ(e) = case x of
+      --          x -> z
+      --           where z = x
+      --          y -> x
+      let subst    = singleSubst (S.unQual z) (S.var x)
+          e        = ["case x of", "  x -> z", "   where z = x", "  y -> z"]
+          expected = ["case x of", "  x -> z", "   where z = x", "  y -> x"]
+      in shouldSubstituteTo subst e expected
+    it "does not rename variables bound later in the where clause"
+      $ runTest
+      $
+      -- σ    = { y ↦ x, z ↦ x }
+      -- e    = case x of
+      --          x_0 -> x
+      --           where x = y
+      --                 y = z
+      -- σ(e) = case x of
+      --          x_0 -> x_0
+      --           where x_0 = y
+      --                 y   = x
+      let subst
+            = substFromList [(S.unQual y, S.var x), (S.unQual z, S.var x)]
+          e
+            = ["case x of", "  x_0 -> x", "   where x = y", "         y = z"]
+          expected = [ "case x of"
+                     , "  x_0 -> x_0"
+                     , "   where x_0 = y"
+                     , "         y   = x"
+                     ]
+      in shouldSubstituteTo subst e expected
+  context "substitution in modules" $ do
+    it "avoids capture of variables shadowed in let expressions"
+      $ runTest
+      $ let m        = ["module A where", "f (x, y) = let x = y in x"]
+            expected = [ "module A where"
+                       , "f a0 = case a0 of"
+                       , "  (a1, a2) -> let x = a2 in x"
+                       ]
+        in m `shouldTransformTo` expected
+    {- TODO The following test should be used instead of the replacement that
+        is not commented out, after the corresponding issue has been fixed.
+    it "avoids capture of variables shadowed in case expressions" $ runTest $
+      let m        = [ "module A where"
+                     , "f (x, y) = case x of"
+                     , "             y -> (x, y)"]
+          expected = [ "module A where"
+                     , "f a0 = case a0 of"
+                     , "  (a1, a2) -> case a1 of"
+                     , "    y -> (a1, y)"]
+      in m `shouldTransformTo` expected -}
+    it "avoids capture of variables shadowed in case expressions"
+      $ runTest
+      $ let m        = [ "module A where"
+                       , "f (x, y) ="
+                       , "  let r = case x of"
+                       , "            y -> (x, y)"
+                       , "  in  r"
+                       ]
+            expected = [ "module A where"
+                       , "f a1 = case a1 of"
+                       , "  (a2, a3) ->"
+                       , "    let r = case a2 of"
+                       , "              a0 -> (a2, a0)"
+                       , "    in  r"
+                       ]
+        in m `shouldTransformTo` expected
```

Comment:
Since these tests test the entire transformation and not just the implementation of substitutions, they are not unit tests. You can move them to the `ApplicationTests` module if you want. `shouldTransformTo` can also be moved back as well.